premium/usage.go

```diff
@@ -0,0 +1,264 @@
+				if time.Since(u.lastUpdateTime) < u.minimumUpdateDuration {
+					// Not enough time since last update
+					continue
+				}
+
+				rowsToUpdate := u.rowsToUpdate.Load()
+				if rowsToUpdate < u.batchLimit {
+					// Not enough rows to update
+					continue
+				}
+				if err := u.updateUsageWithRetryAndBackoff(ctx, rowsToUpdate); err != nil {
+					log.Warn().Err(err).Msg("failed to update usage")
+					continue
+				}
+				u.rowsToUpdate.Add(-rowsToUpdate)
+			case <-flushDuration.C:
+				if time.Since(u.lastUpdateTime) < u.minimumUpdateDuration {
+					// Not enough time since last update
+					continue
+				}
+				rowsToUpdate := u.rowsToUpdate.Load()
+				if rowsToUpdate == 0 {
+					continue
+				}
+				if err := u.updateUsageWithRetryAndBackoff(ctx, rowsToUpdate); err != nil {
+					log.Warn().Err(err).Msg("failed to update usage")
+					continue
+				}
+				u.rowsToUpdate.Add(-rowsToUpdate)
+			case <-u.done:
+				remainingRows := u.rowsToUpdate.Load()
+				if remainingRows != 0 {
+					if err := u.updateUsageWithRetryAndBackoff(ctx, remainingRows); err != nil {
+						u.closeError <- err
+						return
+					}
+					u.rowsToUpdate.Add(-remainingRows)
+				}
+				u.closeError <- nil
+				return
+			}
+		}
+	}()
+	<-started
+}
+
+func (u *BatchUpdater) updateUsageWithRetryAndBackoff(ctx context.Context, numberToUpdate uint32) error {
+	for retry := 0; retry < u.maxRetries; retry++ {
+		queryStartTime := time.Now()
+
+		resp, err := u.apiClient.IncreaseTeamPluginUsageWithResponse(ctx, u.teamName, cqapi.IncreaseTeamPluginUsageJSONRequestBody{
+			RequestId:  uuid.New(),
+			PluginTeam: u.pluginTeam,
+			PluginKind: cqapi.PluginKind(u.pluginKind),
+			PluginName: u.pluginName,
+			Rows:       int(numberToUpdate),
+		})
+		if err != nil {
+			return fmt.Errorf("failed to update usage: %w", err)
+		}
+		if resp.StatusCode() == http.StatusOK {
+			u.lastUpdateTime = time.Now().UTC()
+			return nil
+		}
+
+		retryDuration, err := u.calculateRetryDuration(resp.StatusCode(), resp.HTTPResponse.Header, queryStartTime, retry)
+		if err != nil {
+			return fmt.Errorf("failed to calculate retry duration: %w", err)
+		}
+		if retryDuration > 0 {
+			time.Sleep(retryDuration)
+		}
+	}
+	return fmt.Errorf("failed to update usage: max retries exceeded")
+}
+
+// calculateRetryDuration calculates the duration to sleep relative to the query start time before retrying an update
+func (u *BatchUpdater) calculateRetryDuration(statusCode int, headers http.Header, queryStartTime time.Time, retry int) (time.Duration, error) {
+	if statusCode == http.StatusTooManyRequests {
+		retryAfter := headers.Get("Retry-After")
+		if retryAfter != "" {
+			retryDelay, err := time.ParseDuration(retryAfter + "s")
+			if err != nil {
+				return 0, fmt.Errorf("failed to parse retry-after header: %w", err)
+			}
+			if retryDelay > u.maxWaitTime {
+				return 0, fmt.Errorf("retry-after header exceeds max wait time: %s > %s", retryDelay, u.maxWaitTime)
+			}
```

Comment:
I'm not sure this should be an error. Maybe we should respect whatever delay the server asks us to add?